src/zeroconf/_services/browser.py

```diff
@@ -250,43 +316,115 @@ def _generate_first_next_time(self, now: float_) -> None:
+        if current is not None:
+            # If the expire time is within self._min_time_between_queries_millis
+            # of the current scheduled time avoid churn by not rescheduling
+            if (
+                -self._min_time_between_queries_millis
+                <= expire_time - current.when_millis
+                <= self._min_time_between_queries_millis
+            ):
+                return
+            current.cancelled = True
+        self._schedule(pointer, expire_time)
+
+    def _process_startup_queries(self) -> None:
+        if TYPE_CHECKING:
+            assert self._loop is not None
+        # This is a safety to ensure we stop sending queries if Zeroconf instance
+        # is stopped without the browser being cancelled
+        if self._browser.zc.done:
+            return
+
+        now_millis = current_time_millis()
+
+        # At first we will send 3 queries to get the cache populated
+        self._browser.async_send_ready_queries(True, now_millis, self._browser.types)
+        self._startup_queries_sent += 1
+
+        # Once we finish sending the initial queries we will
+        # switch to a strategy of sending queries only when we
+        # need to refresh records that are about to expire
+        if self._startup_queries_sent >= STARTUP_QUERIES:
+            self._process_ready_types()
+            return
+
+        self._next_run = self._loop.call_later(self._startup_queries_sent**2, self._process_startup_queries)
+
+    def _process_ready_types(self) -> None:
         """Generate a list of ready types that is due and schedule the next time."""
-        if self.millis_to_wait(now):
-            return []
+        if TYPE_CHECKING:
+            assert self._loop is not None
+        # This is a safety to ensure we stop sending queries if Zeroconf instance
+        # is stopped without the browser being cancelled
+        if self._browser.zc.done:
+            return
+
+        now_millis = current_time_millis()
+        # Refresh records that are about to expire (aka
+        # _EXPIRE_REFRESH_TIME_PERCENT which is currently 75% of the TTL) and
+        # with a minimum time between queries of _min_time_between_queries
+        # which defaults to 10s
 
-        ready_types: List[str] = []
+        ready_types: Set[str] = set()
+        next_scheduled: Optional[_ScheduledPTRQuery] = None
+        end_time_millis = now_millis + self._clock_resolution_millis
 
-        for type_, due in self._next_time.items():
-            if due > now:
+        while self._query_heap:
+            query = self._query_heap[0]
+            if query.cancelled:
+                heappop(self._query_heap)
                 continue
+            if query.when_millis > end_time_millis:
+                next_scheduled = query
+                break
+            heappop(self._query_heap)
+            del self._next_scheduled_for_alias[query.alias]
+            ready_types.add(query.name)
```

Comment:
Should we schedule it again in case the rescue fails

Reply:
If the rescue fails try to schedule at 5 pct more.  We should always schedule actually and let it fail